Clarify diagnostic when constant pattern has non-constant conversion to input type

ArcadeMode opened the pull request:
This proposes a fix for #63476 
The newly introduced error message goes for example: ~~`error CS9096: Cannot implicitly convert type 'string' to type 'XName' in constant pattern using non-constant conversion`~~ `error CS9098: Cannot implicitly convert type 'string' to type 'XName' using user-defined conversion in pattern expression`

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=64741) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=64741) before we can accept your contribution.<br/><br/>:x: ArcadeMode [sign now](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=64741)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/roslyn?pullRequest=64741) it.</sub>

AlekseyTs commented:
It looks like there are no tests added or modified.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:
```diff
@@ -542,7 +542,12 @@ private static ExpressionSyntax SkipParensAndNullSuppressions(ExpressionSyntax e
+                var conversion = convertedExpression as BoundConversion;
+                if (conversion != null && conversion.ConversionKind == ConversionKind.ImplicitUserDefined)
+                {
+                    diagnostics.Add(ErrorCode.ERR_NonConstantConversionInConstantPattern, patternExpression.Location, conversion.Operand.Type!, conversion.Type);
```
>conversion.Operand [](http://example.com/codeflow?start=118&length=18)

I think in some cases ```conversion.Operand``` this might not correspond to the syntactic operand for the conversion, Therefore, the type could be surprising and the error message would not make sense because, for example, there is a conversion that supports constant folding between the types.

ArcadeMode replied:
I have updated this to use `expression` instead - this should be the expression being bound and as I understand that does makes sense. Stepping through a few unit tests confirmed my beliefs. This change however, caused errors when the expression was a null literal so I have written a case to show ErrorCode.ERR_ValueCantBeNull when null literals are attempted to be converted to a non-nullable type.

For the following snipped: 
`static bool M1(Span<char> chars) => chars is null;`
This yields the following message:
`// (5,50): error CS0037: Cannot convert null to 'Span<char>' because it is a non-nullable value type` 
instead of 
`// (5,50): error CS0150: A constant value is expected`

Affected test cases were adjusted to this change.

Looking forward to hear your feedback on this change.

RikkiGibson replied:
What kind of conversion is actually used for `Span<char> x = null`?

ArcadeMode replied:
That appears to be an ImplicitReference conversion.

ArcadeMode replied:
> That appears to be an ImplicitReference conversion.

I started doubting the integrity of my recent change with respect to null literals and have taken another look. In the pattern expression the conversion is different, its actually implicit user defined. When there does not exist such a conversion the correct `CS0037: Cannot convert null to 'Span<char>' because it is a non-nullable value type` is already shown so I will proceed to remove the added check for that.

Now I wonder: when an implicit user defined conversion such as `static implicit operator Span<T>(T[]? array)` exists, then the error message in a pattern which attempts to convert null to Span would become `CS9098: Cannot implicitly convert type 'char[]' to type 'Span<char>' using user-defined conversion in pattern expression`. Would this be appropriate?

AlekseyTs commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -7382,4 +7382,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
     <value>A deconstruction variable cannot be declared as a ref local</value>
   </data>
+  <data name="ERR_NonConstantConversionInConstantPattern" xml:space="preserve">
+    <value>Cannot implicitly convert type '{0}' to type '{1}' in constant pattern using non-constant conversion</value>
```
>non-constant conversion [](http://example.com/codeflow?start=88&length=23)

I don't think there is a concept of "non-constant conversion" or "constant conversion" in the language. Since the error is reported specifically for user-defined conversions, we should consider saying exactly that: "Cannot use user-defined conversion for pattern expression." or some thing similar.

ArcadeMode replied:
I have clarified the message as per your recommendation.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:
```diff
@@ -542,7 +542,12 @@ private static ExpressionSyntax SkipParensAndNullSuppressions(ExpressionSyntax e
+                }
+                else if (constantValueOpt == null)
                 {
                     diagnostics.Add(ErrorCode.ERR_ConstantExpected, patternExpression.Location);
```
>diagnostics.Add(ErrorCode.ERR_ConstantExpected, patternExpression.Location); [](http://example.com/codeflow?start=20&length=76)

Is there really a guarantee that we will never get here for a built-in conversion that starts with a constant? What if instead of singling out user defined conversions, we report an error here that would include the ```inputType```. Something like: "A constant value of type 'XName' is expected". To be clear, I am suggesting this as an alternative to the current change, no special handling for user-defined conversions, etc.

ArcadeMode replied:
Judging by Binder_Patterns.cs lines 667 through 669, a comment describes that user-defined conversions do not have a constant value - so I'd say that part is covered. (please correct me if im wrong). 
Now you mention built-in conversions starting with a constant but implementing a non-constant conversion, I have searched but not found any such conversions, admittedly I am not very familar with the internals of roslyn yet so I'll gladly be corrected on this too. 

If any such conversions may be mentioned, I could write up a test case for it and take it from there.

I did find that the implicit conversion from string to ReadOnlySpan<char> is now also possible within the pattern expression and appears to compile just fine, is this expected behavior or should not be allowed like user-defined conversions?

Besides this I do agree that clarifying the message belonging to ERR_ConstantExpected to include the type of constant expected to be an improvement, perhaps another issue would be suitable for this? I'd gladly pick it up.

RikkiGibson replied:
Reply to https://github.com/dotnet/roslyn/pull/64741#discussion_r996202027

Unresolving the comment thread as I wasn't sure @AlekseyTs saw and had the chance to respond. We will want to agree on an approach before moving forward.

> If [constant-to-non-constant conversions] do not exist I believe the current implementation to be clearer to the user.

I think I agree. A user-defined conversion is a like a method call that occurs implicitly based on how an expression is used. If your pattern is a method call `M(a)`, and the compiler says, "you need a constant value", it seems obvious that the issue is a method call does not have a constant value. If that method call is occurring implicitly because it's a conversion operator, then it's not so obvious. With Aleksey's suggestion you can infer that the conversion from the pattern-expression to the pattern input type is not a constant value, but if you are not already familiar with user-defined conversions in the language, this still might be confusing.

RikkiGibson replied:
Have discussed offline. We think that just including the expected type of the constant value, in this case `inputType`, in the diagnostic message will be sufficient to improve the experience here. This will free us up from needing to maintain or expand on the diagnostic messages for these as the behavior of conversions in the language/compiler evolves over time.

Let's change the new error code to something like `ERR_ConstantValueOfTypeExpected`, and the message to `A constant value of type '{0}' is expected`. Thanks!

AlekseyTs commented:
Done with review pass (commit 6).

ArcadeMode commented:
CI fails after updating my fork - seems unrelated to changes in this PR as far as I can tell.

RikkiGibson commented:
Thanks for opening the PR. Will take a look as soon as I am able.

RikkiGibson commented:
Please refrain from force pushing to PRs which are being reviewed. It makes it harder for us to review the revisions incrementally. We can always squash merge your PR (and will unless you strongly object) to keep the history looking nice.

ArcadeMode commented:
Noted, I figured a rebase would omit the unecessary merge commit but I see how it messes up the review process. I will refrain from doing so from now on.

RikkiGibson commented on src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:
```diff
@@ -544,7 +544,21 @@ private static ExpressionSyntax SkipParensAndNullSuppressions(ExpressionSyntax e
+                        }
+                        else
+                        {
+                            diagnostics.Add(ErrorCode.ERR_NonConstantConversionInConstantPattern, patternExpression.Location, expression.Type!, inputType);
```
```suggestion
                            Debug.Assert(expression.Type is not null);
                            diagnostics.Add(ErrorCode.ERR_NonConstantConversionInConstantPattern, patternExpression.Location, expression.Type, inputType);
```

Prefer debug assertions over `!` suppressions.

ArcadeMode commented on src/Compilers/CSharp/Test/Emit2/Semantics/PatternMatchingTests.cs:
```diff
@@ -11784,5 +11784,57 @@ public void Test2(G<object> x2)
+}
+class C
+{
+    static bool M(A a) => a switch { ""implicitA"" => true, _ => false };
```
Hey good catch, I have extended the code to test for ExplicitUserDefined conversions too since its the user defined part that is problematic and generalised the error message to make sense for both implicit and explicit.

ArcadeMode commented:
Hey @RikkiGibson I'd say this is ready for review

RikkiGibson commented:
Holidays are over so will try and give this a review pass when I get the chance. Thanks!

RikkiGibson commented on src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:
```diff
@@ -2150,13 +2150,13 @@ internal enum ErrorCode
         ERR_ImplicitlyTypedDefaultParameter = 9098,
         WRN_OptionalParamValueMismatch = 9099,
         WRN_ParamsArrayInLambdaOnly = 9100,
+        ERR_NonConstantConversionInConstantPattern = 9101,
```
If we keep the approach of checking for user-defined conversions, then let's change this to `ERR_UserDefinedConversionInConstantPattern`.

RikkiGibson commented on src/Compilers/CSharp/Test/Emit2/Semantics/PatternMatchingTests.cs:
```diff
@@ -11784,5 +11786,80 @@ public void Test2(G<object> x2)
+}";
+            CreateCompilationWithSpanAndMemoryExtensions(source, parseOptions: TestOptions.RegularPreview)
+                .VerifyDiagnostics(
+                    // (8,38): error CS9098: Cannot convert type 'string' to type 'A' using user-defined conversion in pattern expression
```
We should also test a subpattern case like `a switch { "implicitA" or { S: { } } => true, ... }`.

RikkiGibson commented:
Will mark as draft for now. Feel free to mark ready for review again once https://github.com/dotnet/roslyn/pull/64741#discussion_r1087299872 is addressed.

ArcadeMode commented:
> Will mark as draft for now. Feel free to mark ready for review again once [#64741 (comment)](https://github.com/dotnet/roslyn/pull/64741#discussion_r1087299872) is addressed.

Been a while! I have addressed the comment, so the PR is ready for another review.

RikkiGibson commented:
/azp run

RikkiGibson commented on src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:
```diff
@@ -557,7 +557,15 @@ private static ExpressionSyntax SkipParensAndNullSuppressions(ExpressionSyntax e
             {
                 if (constantValueOpt == null)
                 {
-                    diagnostics.Add(ErrorCode.ERR_ConstantExpected, patternExpression.Location);
```
I think the suggestion from https://github.com/dotnet/roslyn/pull/64741#discussion_r996202027 was to not introduce any new logic, and change this to just use the new error code you added, passing in the `inputType`.

ArcadeMode replied:
I've removed the check for user defined conversions. However a few tests where the input type was unknown you'd end up with messages like `Constant value of type '?' expected`. Which I dont think improves upon the existing behavior.

For this reason I've now implemented to conditionally return the original `ERR_ConstantExpected` when the inputType is an error symbol. If this is undesirable please let me know.

AlekseyTs commented:
@ArcadeMode Could you please resolve conflicts?

AlekseyTs commented on src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:
```diff
@@ -2150,7 +2150,6 @@ internal enum ErrorCode
         ERR_RefAssignValEscapeWider = 9096,
         WRN_RefAssignValEscapeWider = 9097,
-
         #endregion
```
> #endregion [](http://example.com/codeflow?start=7&length=11)

It looks like an empty line was removed above this directive. Consider restoring it.

ArcadeMode replied:
restored

AlekseyTs commented:
Done with review pass (commit 34). Other than the merge conflicts and minor formatting suggestion, LGTM.

ArcadeMode commented:
> @ArcadeMode Could you please resolve conflicts?

@AlekseyTs resolved and handled formatting suggestion.

RikkiGibson commented:
It looks like some test baselines needs to be updated. https://dev.azure.com/dnceng-public/public/_build/results?buildId=214143&view=ms.vss-test-web.build-test-results-tab

For example:
```
xUnit.net 00:00:24.38]     Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen.SwitchTests.SwitchOnNullableWithNonConstant [FAIL]
  Failed Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen.SwitchTests.SwitchOnNullableWithNonConstant [18 ms]
  Error Message:
   
Expected:
                Diagnostic(ErrorCode.ERR_ConstantExpected, "i").WithLocation(8, 18)
Actual:
                // (8,18): error CS9135: A constant value of type 'int?' is expected
                //             case i:
                Diagnostic(ErrorCode.ERR_ConstantValueOfTypeExpected, "i").WithArguments("int?").WithLocation(8, 18)
Diff:
++>                 Diagnostic(ErrorCode.ERR_ConstantValueOfTypeExpected, "i").WithArguments("int?").WithLocation(8, 18)
-->                 Diagnostic(ErrorCode.ERR_ConstantExpected, "i").WithLocation(8, 18)
Expected: True
Actual:   False
  Stack Trace:
     at Microsoft.CodeAnalysis.DiagnosticExtensions.Verify(IEnumerable`1 actual, DiagnosticDescription[] expected, Boolean errorCodeOnly) in /_/src/Compilers/Test/Core/Diagnostics/DiagnosticExtensions.cs:line 99
   at Microsoft.CodeAnalysis.DiagnosticExtensions.Verify(IEnumerable`1 actual, DiagnosticDescription[] expected) in /_/src/Compilers/Test/Core/Diagnostics/DiagnosticExtensions.cs:line 48
   at Microsoft.CodeAnalysis.DiagnosticExtensions.Verify(ImmutableArray`1 actual, DiagnosticDescription[] expected) in /_/src/Compilers/Test/Core/Diagnostics/DiagnosticExtensions.cs:line 63
   at Microsoft.CodeAnalysis.DiagnosticExtensions.VerifyDiagnostics[TCompilation](TCompilation c, DiagnosticDescription[] expected) in /_/src/Compilers/Test/Core/Diagnostics/DiagnosticExtensions.cs:line 109
   at Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen.SwitchTests.SwitchOnNullableWithNonConstant() in /_/src/Compilers/CSharp/Test/Emit/CodeGen/SwitchTests.cs:line 2625
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

ArcadeMode commented:
I've updated the tests (I hope I got all of them). Oddly enough The failure reported [here](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-roslyn-refs-pull-64741-merge-70dc9bca7b4b4858aa/Microsoft.CodeAnalysis.CSharp.Semantic.UnitTests_29/1/console.d3360de9.log?helixlogtype=result) I couldn't reproduce locally as the baseline was already set to expect the new error code.. any clue what might be going on there?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/SwitchTests.cs:
```diff
@@ -2623,10 +2623,10 @@ static void Main()
             var compilation = base.CreateCSharpCompilation(text);
             compilation.VerifyDiagnostics(
-                // (8,18): error CS0150: A constant value is expected
+                // (8,18): error CS9135: A constant value of type 'int?' is expected
```
>// (8,18): error CS9135: A constant value of type 'int?' is expected [](http://example.com/codeflow?start=16&length=68)

This is actually an interesting scenario. I think we should strip the nullability from value types used in the message. We have ```StrippedType``` helper for this.

ArcadeMode replied:
Done!

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_ISwitchOperation.cs:
```diff
@@ -2076,9 +2076,9 @@ void M(bool result, dynamic input, dynamic other)
 ";
             var expectedDiagnostics = new[] {
-                // file.cs(8,18): error CS0150: A constant value is expected
+                // (8,18): error CS9135: A constant value of type 'dynamic' is expected
```
> // (8,18): error CS9135: A constant value of type 'dynamic' is expected [](http://example.com/codeflow?start=15&length=72)

This one is another interesting scenario. I think we should stick to the old message for dynamic, object and System.ValueType. Please add tests for these scenarios that are not currently covered.

ArcadeMode replied:
Done!

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -115790,13 +115790,13 @@ void M0<T>(T x0)
 
             CreateCompilation(source, options: WithNullableEnable()).VerifyDiagnostics(
-                // (6,19): error CS0150: A constant value is expected
+                // (6,19): error CS9135: A constant value of type 'T' is expected
```
>// (6,19): error CS9135: A constant value of type 'T' is expected [](http://example.com/codeflow?start=16&length=65)

For a generic type parameter should probably stick to the old message too.

ArcadeMode replied:
Done!

AlekseyTs commented:
Done with review pass (commit 37)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit2/FlowAnalysis/FlowDiagnosticTests.cs:
```diff
@@ -2759,5 +2759,27 @@ struct Empty
+        [InlineData("object", "new {}")]
+        [InlineData("dynamic", "new {}")]
+        [InlineData("System.ValueType", "((int)x0)/2")]
+        public void ConstantsExpectedInPatternExpression(string type, string expression)
```
>ConstantsExpectedInPatternExpression [](http://example.com/codeflow?start=20&length=36)

The placement of this test is somewhat unexpected. Consider moving it to a file/type dedicated to patterns.

AlekseyTs commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_ISwitchOperation.cs:
```diff
@@ -1813,9 +1813,9 @@ void M(bool result, int? input, int? other)
 ";
             var expectedDiagnostics = new[] {
-                // file.cs(8,18): error CS0150: A constant value is expected
+                // (8,18): error CS9135: A constant value of type 'int?' is expected
```
>? [](http://example.com/codeflow?start=70&length=1)

Should '?' be removed?

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:
```diff
@@ -77480,13 +77480,13 @@ static void G(object? o)
             comp = CreateCompilation(new[] { source }, options: WithNullableEnable());
             comp.VerifyDiagnostics(
-                // (6,18): warning CS8974: Converting method group 'F' to non-delegate type 'object'. Did you intend to invoke the method?
+                // 0.cs(6,18): warning CS8974: Converting method group 'F' to non-delegate type 'object'. Did you intend to invoke the method?
```
> 0.cs [](http://example.com/codeflow?start=18&length=5)

It looks like there are no meaningful changes in this file. I recommend restoring it to its original state.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesVsPatterns.cs:
```diff
@@ -174,13 +174,13 @@ void Test(object? x)
 ");
             c.VerifyDiagnostics(
-                // (7,18): error CS0150: A constant value is expected
+                // 0.cs(7,18): error CS0150: A constant value is expected
```
> 0.cs [](http://example.com/codeflow?start=18&length=5)

It looks like there are no meaningful changes in this file. I recommend restoring it to its original state.

AlekseyTs commented:
Done with review pass (commit 40)

ArcadeMode commented:
addressed feedback

AlekseyTs reviewed (approved):
LGTM (commit 43)

AlekseyTs commented:
@RikkiGibson For the second review

RikkiGibson commented:
Thanks for the contribution @ArcadeMode!
